src/device/mmio.rs

```diff
@@ -0,0 +1,355 @@
+                    0xfc => self.config_generation().into(),
+                    _ => {
+                        warn!("unknown virtio mmio register read: 0x{:x}", offset);
+                        return;
+                    }
+                };
+                // This cannot panic, because we checked that `data.len() == 4`.
+                data.copy_from_slice(v.to_le_bytes().as_ref());
+            }
+
+            // TODO: The standard specifies that configuration space size is device specific, so
+            // we might want to express that via the trait instead of hard coding the current
+            // arbitrary ceiling.
+            // It's ok to use `as` here because `offset` always fits into an `usize` in this case.
+            0x100..=0xfff => self.read_config(offset as usize - 0x100, data),
+            _ => {
+                warn!(
+                    "invalid virtio mmio read: 0x{:x}:0x{:x}",
+                    offset,
+                    data.len()
+                );
+            }
+        };
+    }
+
+    /// Handle a driver write operation to the MMIO space of the device.
+    fn write(&mut self, offset: u64, data: &[u8]) {
+        match offset {
+            // The standard specifies that accesses to configuration registers are 32-bit wide.
+            0x00..=0xff if data.len() == 4 => {
+                // The `try_into` below attempts to convert `data` to a `[u8; 4]`, which
+                // always succeeds because we previously checked that `data.len() == 4`.
+                let v = u32::from_le_bytes(data.try_into().unwrap());
+                match offset {
+                    0x14 => self.set_device_features_select(v),
+                    0x20 => {
+                        if self.check_device_status(
+                            status::DRIVER,
+                            status::FEATURES_OK | status::FAILED,
+                        ) {
+                            self.set_driver_features(self.driver_features_select(), v);
+                        } else {
+                            warn!(
+                                "ack virtio features in invalid state 0x{:x}",
+                                self.device_status()
+                            );
+                        }
+                    }
+                    0x24 => self.set_driver_features_select(v),
+
+                    // TODO: add warnings or signal some sort of event (depending on how we end up
+                    // implementing logging and metrics) for values that do not actually fit the
+                    // data type specified by the virtio standard (we simply use `as` conversion
+                    // for now).
+                    0x30 => self.set_queue_select(v as u16),
+                    0x38 => update_queue_field(self, |q| q.size = v as u16),
+                    0x44 => update_queue_field(self, |q| q.ready = v == 1),
+                    0x50 => self.queue_notify(v),
```

Comment:
This is also rather easily testable if you override `fn queue_notify` for `Dummy` and force a write here.